internal/oidc/auth/auth_handler_test.go

```diff
@@ -145,11 +172,33 @@ func TestAuthorizationEndpoint(t *testing.T) {
 	upstreamAuthURL, err := url.Parse("https://some-upstream-idp:8443/auth")
 	require.NoError(t, err)
 
-	upstreamOIDCIdentityProvider := oidctestutil.TestUpstreamOIDCIdentityProvider{
-		Name:             "some-oidc-idp",
-		ClientID:         "some-client-id",
-		AuthorizationURL: *upstreamAuthURL,
-		Scopes:           []string{"scope1", "scope2"}, // the scopes to request when starting the upstream authorization flow
+	upstreamOIDCIdentityProvider := func() *oidctestutil.TestUpstreamOIDCIdentityProvider {
+		return oidctestutil.NewTestUpstreamOIDCIdentityProviderBuilder().
+			WithName(oidcUpstreamName).
+			WithClientID("some-client-id").
+			WithAuthorizationURL(*upstreamAuthURL).
+			WithScopes([]string{"scope1", "scope2"}). // the scopes to request when starting the upstream authorization flow
+			WithAllowPasswordGrant(false).
+			WithPasswordGrantError(errors.New("should not have used password grant on this instance")).
+			Build()
+	}
+
+	passwordGrantUpstreamOIDCIdentityProviderBuilder := func() *oidctestutil.TestUpstreamOIDCIdentityProviderBuilder {
+		return oidctestutil.NewTestUpstreamOIDCIdentityProviderBuilder().
+			WithName(oidcPasswordGrantUpstreamName).
+			WithClientID("some-client-id").
+			WithAuthorizationURL(*upstreamAuthURL).
+			WithScopes([]string{"scope1", "scope2"}). // the scopes to request when starting the upstream authorization flow
+			WithAllowPasswordGrant(false).
+			WithUsernameClaim(oidcUpstreamUsernameClaim).
+			WithGroupsClaim(oidcUpstreamGroupsClaim).
+			WithIDTokenClaim("iss", oidcUpstreamIssuer).
+			WithIDTokenClaim("sub", oidcUpstreamSubject).
+			WithIDTokenClaim(oidcUpstreamUsernameClaim, oidcUpstreamUsername).
+			WithIDTokenClaim(oidcUpstreamGroupsClaim, oidcUpstreamGroupMembership).
+			WithIDTokenClaim("other-claim", "should be ignored").
+			WithAllowPasswordGrant(true).
+			WithUpstreamAuthcodeExchangeError(errors.New("should not have tried to exchange upstream authcode on this instance"))
 	}
```

Comment:
I agree that over use of closures can make for confusing code, and I'm sure that I'm guilty of overdoing it in some of our older tests.

However, in this case every variable used inside the closures is either `const` or is effectively `const` (e.g. `upstreamAuthURL` which would be a const except that it needs to be the result of `url.Parse`), and the functions simply return new structs. Using a simple closure as a test helper to create a struct containing fake test data seems reasonable and helps keep each test in the test table brief and expressive.

Reply:
> keep each test in the test table brief and expressive

I think about this differently - I want the table test to be verbose and straightforward, because that is the general nature of table tests - repetition is both fine and generally preferred.  So even in this case, a regular `func passwordGrantUpstreamOIDCIdentityProviderBuilder(... params here...)` is preferred over the closure.  But a conversation for another day.